### PR TITLE
owner(ticdc): fix protocol incompatibility in sink scheme during owner bootstrap

### LINKED ISSUE #6459: v5.1.4 upgrade to v6.2.0, cdc task status turn to error: CDC:ErrSinkURIInvalid
### What did you do?

1. install v5.1.4 tidb cluster
2. init cdc task
3. upgrade to v6.2.0
4. check cdc task status

### What did you expect to see?

task status is normal

### What did you see instead?

[2022/07/25 14:18:47.495 +08:00] [ERROR] [changefeed.go:204] ["an error occurred in Owner"] [namespace=default] [changefeed=simple-cdc-task] [error="[CDC:ErrSinkURIInvalid]sink uri invalid%!(EXTRA string=protocol cannot be configured when using tidb scheme)"] [errorVerbose="[CDC:ErrSinkURIInvalid]sink uri invalid%!(EXTRA string=protocol cannot be configured when using tidb scheme)\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/config.(*SinkConfig).applyParameter\n\tgithub.com/pingcap/tiflow/pkg/config/sink.go:178\ngithub.com/pingcap/tiflow/pkg/config.(*SinkConfig).validateAndAdjust\n\tgithub.com/pingcap/tiflow/pkg/config/sink.go:90\ngithub.com/pingcap/tiflow/pkg/config.(*ReplicaConfig).ValidateAndAdjust\n\tgithub.com/pingcap/tiflow/pkg/config/replica_config.go:120\ngithub.com/pingcap/tiflow/cdc/sink.New\n\tgithub.com/pingcap/tiflow/cdc/sink/sink.go:163\ngithub.com/pingcap/tiflow/cdc/owner.ddlSinkInitializer\n\tgithub.com/pingcap/tiflow/cdc/owner/ddl_sink.go:106\ngithub.com/pingcap/tiflow/cdc/owner.(*ddlSinkImpl).run.func1\n\tgithub.com/pingcap/tiflow/cdc/owner/ddl_sink.go:139\nruntime.goexit\n\truntime/asm_amd64.s:1571"]
[root@cdc-0 bin]# ./cdc cli changefeed list --pd=http://pd-peer:12379/
[
  {
    "id": "simple-cdc-task",
    "namespace": "default",
    "summary": {
      "state": "error",
      "tso": 434823975864631309,
      "checkpoint": "2022-07-25 12:03:51.759",
      "error": {
        "addr": "cdc-peer:8300",
        "code": "CDC:ErrSinkURIInvalid",
        "message": "[CDC:ErrSinkURIInvalid]sink uri invalid%!(EXTRA string=protocol cannot be configured when using tidb scheme)"
      }
    }
  }
]

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
| Release Version: v6.2.0
Edition: Community
Git Commit Hash: https://github.com/pingcap/tidb/commit/0ea937e56adac89ea7df1ce5406c7dc20a3f83ae
Git Branch: heads/refs/tags/v6.2.0
UTC Build Time: 2022-07-23 10:09:51
GoVersion: go1.18.2
```

Upstream TiKV version (execute `tikv-server --version`):

```console
(paste TiKV version here)
```

TiCDC version (execute `cdc version`):

```console
[root@cdc-0 bin]# ./cdc version
Release Version: v6.2.0
Git Commit Hash: 934824b8c065e1a14c840e5927fac48b83289576
Git Branch: heads/refs/tags/v6.2.0
UTC Build Time: 2022-07-22 03:06:00
Go Version: go version go1.18.2 linux/amd64
Failpoint Build: false
[root@cdc-0 bin]# 

```

### PR DESCRIPTION
**Pull request by CharlesCheung96**
### What problem does this PR solve?

Issue Number: close #6459

### What is changed and how it works?
1. Log warn message when protocol is `default` and scheme is `mysql or tidb`.
2. Fix protocol incompatibility in sink scheme during owner bootstrap

#### Tests 

 - Unit test
 - Manual Test
1. start a v5.1.4 cdc server 
3. create 2 changefeeds
```shell
./cdc cli changefeed create --sink-uri="mysql://root@127.0.0.1:61551/?time-zone=" -c "test-1" 
./cdc cli changefeed create --sink-uri="mysql://root@127.0.0.1:61551/?time-zone=&protocol=default" -c "test-2" 
```
image
image

4. exit v5.1.4 cdc server, start fixed cdc server
5. query the 2 changefeeds above
image
image
6. list changefeed
image

### Release note 

```release-note
`None`.
```

### CODE REVIEW

**Review by sdojjy — APPROVED** [review]

File: cdc/model/changefeed_test.go [hunk 1]
```diff
@@ -223,10 +223,12 @@ func TestFixStateIncompatible(t *testing.T) {
...
 	t.Parallel()
 
+	emptyProtocolStr := ""
 	// Test to fix incompatible protocols.
 	configTestCases := []struct {
-		info             *ChangeFeedInfo
-		expectedProtocol config.Protocol
+		info                *ChangeFeedInfo
+		expectedProtocol    config.Protocol
+		expectedProtocolStr *string
```
Comment (asddongmen) [inline cdc/model/changefeed_test.go:231]: Why do we need to use a pointer here?

File: cdc/model/changefeed.go [hunk 2]
```diff
@@ -351,6 +351,38 @@ func (info *ChangeFeedInfo) fixSinkProtocol() {
...
+		if newProtocolStr == "" {
+			rawQuery.Del(config.ProtocolKey)
+		} else {
+			rawQuery.Set(config.ProtocolKey, newProtocolStr)
+		}
+		oldRawQuery := sinkURIParsed.RawQuery
+		newRawQuery := rawQuery.Encode()
+		sinkURIParsed.RawQuery = newRawQuery
+		fixedSinkURI := sinkURIParsed.String()
+		log.Info("handle incompatible protocol from sink URI",
```
Comment (asddongmen) [inline cdc/model/changefeed.go:364]: Can we just to print the `oldSinkURI` and `newSinkURI` ?

**Comment by asddongmen** [comment]
=== Failed

[2022-07-25T13:32:29.549Z] === FAIL: cdc/model TestFixSinkProtocol (0.00s)

[2022-07-25T13:32:29.549Z]     changefeed_test.go:587: 

[2022-07-25T13:32:29.549Z]         	Error Trace:	changefeed_test.go:587

[2022-07-25T13:32:29.549Z]         	Error:      	Expected nil, but got: [CDC:ErrMQSinkUnknownProtocol]unknown '' protocol for Message Queue sink

[2022-07-25T13:32:29.549Z]         	Test:       	TestFixSinkProtocol

**Review by asddongmen — APPROVED** [review]

File: cdc/model/changefeed.go [hunk 3]
```diff
@@ -351,6 +351,44 @@ func (info *ChangeFeedInfo) fixSinkProtocol() {
...
+			rawQuery.Del(config.ProtocolKey)
+		} else {
+			rawQuery.Set(config.ProtocolKey, newProtocolStr)
+		}
+		oldSinkURI := sinkURIParsed.String()
+		sinkURIParsed.RawQuery = rawQuery.Encode()
+		info.SinkURI = sinkURIParsed.String()
+		if oldSinkURI != info.SinkURI {
+			log.Info("handle incompatible protocol from sink URI",
+				zap.String("oldSinkURI", oldSinkURI),
```
Comment (hi-rustin) [inline cdc/model/changefeed.go:365]: Does it print password and user information?

File: cdc/model/changefeed.go [hunk 4]
```diff
@@ -351,6 +351,44 @@ func (info *ChangeFeedInfo) fixSinkProtocol() {
 	rawQuery := sinkURIParsed.Query()
 	protocolStr := rawQuery.Get(config.ProtocolKey)
 
+	fixSinkURI := func(newProtocolStr string) {
+		if newProtocolStr == "" {
```
Comment (hi-rustin) [inline cdc/model/changefeed.go:355]: Please split this operation or add a comment for it.

File: cdc/model/changefeed.go [hunk 5]
```diff
@@ -351,6 +351,44 @@ func (info *ChangeFeedInfo) fixSinkProtocol() {
...
+				zap.String("fixedProtocol", info.Config.Sink.Protocol))
+		}
+	}
+
+	// fix mysql sink
+	scheme := sinkURIParsed.Scheme
+	if !config.IsMqScheme(scheme) {
+		if protocolStr != "" || info.Config.Sink.Protocol != "" {
+			log.Warn("sink URI or sink config contains protocol, but scheme is not mq",
+				zap.String("sinkURI", info.SinkURI),
```
Comment (hi-rustin) [inline cdc/model/changefeed.go:382]: Does it print password and user information?

File: pkg/config/sink.go [hunk 6]
```diff
@@ -168,15 +168,18 @@ func (s *SinkConfig) applyParameter(sinkURI *url.URL) error {
 	}
 
 	// validate that protocol is compatible with the scheme
-	if isMqScheme(sinkURI.Scheme) {
+	if IsMqScheme(sinkURI.Scheme) {
```
Comment (hi-rustin) [inline pkg/config/sink.go:171]: Please refine this comment to start with `IsMqScheme`.

File: pkg/config/sink.go [hunk 7]
```diff
@@ -168,15 +168,18 @@ func (s *SinkConfig) applyParameter(sinkURI *url.URL) error {
...
 		var protocol Protocol
 		err := protocol.FromString(s.Protocol)
 		if err != nil {
 			return err
 		}
+	} else if s.Protocol == "default" {
```
Comment (hi-rustin) [inline pkg/config/sink.go:177]: Why we need this?

**Comment by hi-rustin** [comment]
Please rewrite the PR title. Thanks!

File: pkg/config/sink.go (hunk 7 above)
Comment (CharlesCheung96) [reply pkg/config/sink.go:177]: To ease restrictions appropriately.

File: pkg/config/sink.go (hunk 6 above)
Comment (CharlesCheung96) [reply pkg/config/sink.go:171]: Ref: https://github.com/pingcap/tiflow/pull/6461/files#diff-1aa9ba3321cb0edb368a32b5084da34322b0f9d6a600019beb724ab63bc5e071R191

File: pkg/config/sink.go (hunk 7 above)
Comment (hi-rustin) [reply pkg/config/sink.go:177]: But shouldn't it be fixed right at startup?

**Review by hi-rustin — COMMENTED** [review]
rest LGTM

**Review by hi-rustin — APPROVED** [review]
Thanks you all for working on this. I will test it.

**Comment by hi-rustin** [comment]
/hold

I will test it.

**Comment by hi-rustin** [comment]
I tested it:
1. start ticdc: version 5.1.4
1. ./cdc cli changefeed create --sink-uri="mysql://root:12346@127.0.0.1:3306/" --config=changefeed.toml
```toml
[sink]
protocol="maxwell"
```
```log
Create changefeed successfully!
ID: 4d8a7e05-721b-470b-924b-16b77150dfdb
Info: {"sink-uri":"mysql://root@localhost:3306/","opts":{"_changefeed_id":"cli-verify"},"create-time":"2022-07-26T16:40:23.171603+08:00","start-ts":434850974434066433,"target-ts":0,"admin-job-type":0,"sort-engine":"unified","sort-dir":"","config":{"case-sensitive":true,"enable-old-value":true,"force-replicate":false,"check-gc-safe-point":true,"filter":{"rules":["*.*"],"ignore-txn-start-ts":null},"mounter":{"worker-num":16},"sink":{"dispatchers":null,"protocol":"maxwell"},"cyclic-replication":{"enable":false,"replica-id":0,"filter-replica-ids":null,"id-buckets":0,"sync-ddl":false},"scheduler":{"type":"table-number","polling-time":-1}},"state":"normal","error":null,"sync-point-enabled":false,"sync-point-interval":600000000000,"creator-version":"v5.1.4"}
```
2. start ticdc: version pr#6461
```log
[2022/07/26 16:44:12.703 +08:00] [WARN] [changefeed.go:372] ["sink URI or sink config contains protocol, but scheme is not mq"] [sinkURI=mysql://root:xxxx@localhost:3306/] [protocol=] [sinkConfig="{\"dispatchers\":null,\"protocol\":\"maxwell\",\"column-selectors\":null,\"schema-registry\":\"\",\"transaction-atomicity\":\"\"}"]
[2022/07/26 16:44:12.703 +08:00] [INFO] [changefeed.go:357] ["handle incompatible protocol from sink URI"] [oldUriQuery=] [fixedUriQuery=]
[2022/07/26 16:44:12.704 +08:00] [INFO] [changefeed.go:295] ["Fix incompatibility changefeed sink protocol completed"] [changefeed="{\"upstream-id\":7124592497092167171,\"namespace\":\"default\",\"changefeed-id\":\"\",\"sink-uri\":\"mysql://root:xxxx@localhost:3306/\",\"create-time\":\"2022-07-26T16:40:23.171603+08:00\",\"start-ts\":434850974434066433,\"target-ts\":0,\"admin-job-type\":0,\"sort-engine\":\"unified\",\"sort-dir\":\"\",\"config\":{\"case-sensitive\":true,\"enable-old-value\":true,\"force-replicate\":false,\"check-gc-safe-point\":true,\"filter\":{\"rules\":[\"*.*\"],\"ignore-txn-start-ts\":null,\"event-filters\":null},\"mounter\":{\"worker-num\":16},\"sink\":{\"dispatchers\":null,\"protocol\":\"\",\"column-selectors\":null,\"schema-registry\":\"\",\"transaction-atomicity\":\"\"},\"consistent\":null},\"state\":\"normal\",\"error\":null,\"sync-point-enabled\":false,\"sync-point-interval\":600000000000,\"creator-version\":\"v5.1.4\"}"]
```
It works well.

But when I try to update it. I will be an error same as https://github.com/pingcap/tiflow/issues/6336
```shell
 ./cdc cli changefeed update -c 4d8a7e05-721b-470b-924b-16b77150dfdb --config=changefeed.toml
```
```log
./cdc cli changefeed update -c 4d8a7e05-721b-470b-924b-16b77150dfdb --sink-uri="mysql://root:12346@127.0.0.1:3306/?protocol=craft"
Diff of changefeed config:
{Type:update Path:[SinkURI] From:mysql://root@localhost:3306/ To:mysql://root:12346@127.0.0.1:3306/?protocol=craft}
Could you agree to apply changes above to changefeed [Y/N]
Y
Error: invalid character 'p' after top-level value
```

**Comment by hi-rustin** [comment]
I tested another case:
1. start ticdc: version 5.1.4
2. ./cdc cli changefeed create  --sink-uri="mysql://root@localhost:3306/?protocol=random"
```log
Create changefeed successfully!
ID: abb5ce8d-4a51-4710-97d1-6c047985124b
Info: {"sink-uri":"mysql://root@localhost:3306/?protocol=random","opts":{"_changefeed_id":"cli-verify"},"create-time":"2022-07-26T17:05:10.454415+08:00","start-ts":434851364318216193,"target-ts":0,"admin-job-type":0,"sort-engine":"unified","sort-dir":"","config":{"case-sensitive":true,"enable-old-value":true,"force-replicate":false,"check-gc-safe-point":true,"filter":{"rules":["*.*"],"ignore-txn-start-ts":null},"mounter":{"worker-num":16},"sink":{"dispatchers":null,"protocol":"default"},"cyclic-replication":{"enable":false,"replica-id":0,"filter-replica-ids":null,"id-buckets":0,"sync-ddl":false},"scheduler":{"type":"table-number","polling-time":-1}},"state":"normal","error":null,"sync-point-enabled":false,"sync-point-interval":600000000000,"creator-version":"v5.1.4"}
```
2. start ticdc: version pr#6461
```log
[2022/07/26 17:06:17.516 +08:00] [WARN] [changefeed.go:372] ["sink URI or sink config contains protocol, but scheme is not mq"] [sinkURI="mysql://root:xxxx@localhost:3306/?protocol=random"] [protocol=random] [sinkConfig="{\"dispatchers\":null,\"protocol\":\"default\",\"column-selectors\":null,\"schema-registry\":\"\",\"transaction-atomicity\":\"\"}"]
[2022/07/26 17:06:17.516 +08:00] [INFO] [changefeed.go:357] ["handle incompatible protocol from sink URI"] [oldUriQuery="protocol=random"] [fixedUriQuery=]
[2022/07/26 17:06:17.516 +08:00] [INFO] [changefeed.go:295] ["Fix incompatibility changefeed sink protocol completed"] [changefeed="{\"upstream-id\":7124604087815269339,\"namespace\":\"default\",\"changefeed-id\":\"\",\"sink-uri\":\"mysql://root:xxxx@localhost:3306/\",\"create-time\":\"2022-07-26T17:05:10.454415+08:00\",\"start-ts\":434851364318216193,\"target-ts\":0,\"admin-job-type\":0,\"sort-engine\":\"unified\",\"sort-dir\":\"\",\"config\":{\"case-sensitive\":true,\"enable-old-value\":true,\"force-replicate\":false,\"check-gc-safe-point\":true,\"filter\":{\"rules\":[\"*.*\"],\"ignore-txn-start-ts\":null,\"event-filters\":null},\"mounter\":{\"worker-num\":16},\"sink\":{\"dispatchers\":null,\"protocol\":\"\",\"column-selectors\":null,\"schema-registry\":\"\",\"transaction-atomicity\":\"\"},\"consistent\":null},\"state\":\"normal\",\"error\":null,\"sync-point-enabled\":false,\"sync-point-interval\":600000000000,\"creator-version\":\"v5.1.4\"}"]
```

**Comment by hi-rustin** [comment]
I tested another case:
1. start ticdc: version 5.1.4
2. ./cdc cli changefeed create  --sink-uri="mysql://root@localhost:3306/?protocol=random" --config=changefeed.toml
```toml
[sink]
protocol="craft"
```
```log
Create changefeed successfully!
ID: 0d9245eb-15f8-4aa2-b3a4-3e2ba0fdf7cb
Info: {"sink-uri":"mysql://root@localhost:3306/?protocol=random","opts":{"_changefeed_id":"cli-verify"},"create-time":"2022-07-26T17:11:07.657025+08:00","start-ts":434851457964703745,"target-ts":0,"admin-job-type":0,"sort-engine":"unified","sort-dir":"","config":{"case-sensitive":true,"enable-old-value":true,"force-replicate":false,"check-gc-safe-point":true,"filter":{"rules":["*.*"],"ignore-txn-start-ts":null},"mounter":{"worker-num":16},"sink":{"dispatchers":null,"protocol":"craft"},"cyclic-replication":{"enable":false,"replica-id":0,"filter-replica-ids":null,"id-buckets":0,"sync-ddl":false},"scheduler":{"type":"table-number","polling-time":-1}},"state":"normal","error":null,"sync-point-enabled":false,"sync-point-interval":600000000000,"creator-version":"v5.1.4"}
```
2. start ticdc: version pr#6461
```log
[2022/07/26 17:06:17.516 +08:00] [WARN] [changefeed.go:372] ["sink URI or sink config contains protocol, but scheme is not mq"] [sinkURI="mysql://root:xxxx@localhost:3306/?protocol=random"] [protocol=random] [sinkConfig="{\"dispatchers\":null,\"protocol\":\"default\",\"column-selectors\":null,\"schema-registry\":\"\",\"transaction-atomicity\":\"\"}"]
[2022/07/26 17:06:17.516 +08:00] [INFO] [changefeed.go:357] ["handle incompatible protocol from sink URI"] [oldUriQuery="protocol=random"] [fixedUriQuery=]
[2022/07/26 17:06:17.516 +08:00] [INFO] [changefeed.go:295] ["Fix incompatibility changefeed sink protocol completed"] [changefeed="{\"upstream-id\":7124604087815269339,\"namespace\":\"default\",\"changefeed-id\":\"\",\"sink-uri\":\"mysql://root:xxxx@localhost:3306/\",\"create-time\":\"2022-07-26T17:05:10.454415+08:00\",\"start-ts\":434851364318216193,\"target-ts\":0,\"admin-job-type\":0,\"sort-engine\":\"unified\",\"sort-dir\":\"\",\"config\":{\"case-sensitive\":true,\"enable-old-value\":true,\"force-replicate\":false,\"check-gc-safe-point\":true,\"filter\":{\"rules\":[\"*.*\"],\"ignore-txn-start-ts\":null,\"event-filters\":null},\"mounter\":{\"worker-num\":16},\"sink\":{\"dispatchers\":null,\"protocol\":\"\",\"column-selectors\":null,\"schema-registry\":\"\",\"transaction-atomicity\":\"\"},\"consistent\":null},\"state\":\"normal\",\"error\":null,\"sync-point-enabled\":false,\"sync-point-interval\":600000000000,\"creator-version\":\"v5.1.4\"}"]
```

**Comment by hi-rustin** [comment]
Conclusion:
1. This fix works well.
2. There is a problem with the logs, they do not print clear changes (when it is tidb protocol). I would like to print only the protocol changes.

@CharlesCheung96 @asddongmen

**Pull request merged**